search/src/search.rs

```diff
@@ -181,37 +190,63 @@ impl Search {
         Ok(self.highlight(snippet))
     }
 
-    fn get_fuzzy_preview(&self, term: &str, distance: &u8, body: String) -> Result<String> {
-        let pindex: HashMap<&str, Vec<(u32, u32)>> = HashMap::from([
-            ("flooding", vec![(2, 10)])
-        ]);
-        let (matched_term, start, end) = self.get_fuzzy_match(term, distance, pindex);
+    fn get_fuzzy_preview(
+        &self,
+        path: &str,
+        term: &str,
+        distance: &u8,
+        body: &str,
+    ) -> Result<String> {
+        let pindex: PageIndex = self
+            .index
+            .page_index(path)
+            .map_err(|_| SearchError("".to_string()))?;
+        let (matched_term, start, end) = self
+            .get_fuzzy_match(term, distance, pindex)
+            .map_err(|_| SearchError("".to_string()))?;
         // Another safe way to get substrings using char_indices
-        let start = body.char_indices().nth(start as usize).unwrap_or((0, ' ')).0;
-        let end = body.char_indices().nth(end as usize).unwrap_or((body.len()-1, ' ')).0;
-        let substring = &body[start..end];
-        Ok(substring.to_string().replace(&matched_term, &format!("**{}**", term)))
+        let start = body
+            .char_indices()
+            .nth(start.saturating_sub(20) as usize)
+            .unwrap_or((0, ' '))
+            .0;
+        let end = body
+            .char_indices()
+            .nth((end + 20) as usize)
+            .unwrap_or((body.len() - 1, ' '))
+            .0;
+        let substring = &format!("...{}...", &body[start..end]);
+        let substring = substring
+            .to_string()
+            .replace(&matched_term, &format!("**{}**", matched_term));
+        Ok(substring.replace('\n', " "))
     }
 
-    fn get_fuzzy_match(&self, term: &str, distance: &u8, pindex: HashMap<&str, Vec<(u32, u32)>>) -> (String, u32, u32) {
+    fn get_fuzzy_match(
+        &self,
+        term: &str,
+        distance: &u8,
+        pindex: PageIndex,
+    ) -> Result<(String, u32, u32)> {
         if pindex.contains_key(term) {
             let (start, end) = pindex.get(term).unwrap().first().unwrap();
-            return (term.to_string(), *start, *end);
+            Ok((term.to_string(), *start, *end))
         } else {
             let mut cur: (String, u32, u32) = ("".to_string(), 0, 0);
             let mut min_dist: usize = usize::MAX;
             for (word, matches) in pindex {
-                let dist: usize = levenshtein(term, word); 
+                let dist: usize = levenshtein(term, &word);
+                let dist = if word.contains(term) { 1 } else { dist };
```

Comment:
Why do we overwrite the value in `dist` with `1` in this case?

Reply:
The idea is that f.e. "Soledad" should be found when searching for "Sole". However, if you have a good reason for not doing so I'd also be fine.

Reply:
I did however change the syntax to a onliner